lib/segment/src/payload_storage/query_checker.rs

```diff
@@ -100,22 +106,62 @@ where
     check_filter(&checker, query)
 }
 
-pub fn check_is_empty_condition(is_empty: &IsEmptyCondition, payload: &Payload) -> bool {
+pub fn check_is_empty_condition(
+    is_empty: &IsEmptyCondition,
+    payload: &impl PayloadContainer,
+) -> bool {
     payload.get_value(&is_empty.is_empty.key).check_is_empty()
 }
 
-pub fn check_is_null_condition(is_null: &IsNullCondition, payload: &Payload) -> bool {
+pub fn check_is_null_condition(is_null: &IsNullCondition, payload: &impl PayloadContainer) -> bool {
     payload.get_value(&is_null.is_null.key).check_is_null()
 }
 
-pub fn check_field_condition(field_condition: &FieldCondition, payload: &Payload) -> bool {
+pub fn check_field_condition<R>(
+    field_condition: &FieldCondition,
+    payload: &impl PayloadContainer,
+    field_indexes: &HashMap<PayloadKeyType, R>,
+) -> bool
+where
+    R: AsRef<Vec<FieldIndex>>,
+{
     let field_values = payload.get_value(&field_condition.key);
-
-    let mut res = false;
-    for p in field_values {
-        res |= field_condition.check(p);
+    let field_indexes = field_indexes.get(&field_condition.key);
+
+    // This covers a case, when a field index affects the result of the condition.
+    if let Some(field_indexes) = field_indexes {
+        for p in field_values {
+            let mut index_checked = false;
+            for index in field_indexes.as_ref() {
+                if let Some(index_check_res) = index.check_condition(field_condition, p) {
+                    if index_check_res {
+                        // If at least one object matches the condition, we can return true
+                        return true;
+                    }
+                    index_checked = true;
+                    // If index check of the condition returned something, we don't need to check
+                    // other indexes
+                    break;
+                }
+            }
```

Comment:
Different approach without `break`, not sure which is better:

```suggestion
            if let Some(index_check_res) = field_indexes
                .as_ref()
                .iter()
                .filter_map(|index| index.check_condition(field_condition, p))
                .next()
            {
                // If at least one object matches the condition, we can return true
                if index_check_res {
                    return true;
                }
                // If index check of the condition returned something, we don't need to check
                // other indexes
                index_checked = true;
            }
```